### ISSUE TITLE
useNavigate hook causes waste rendering

### ISSUE THREAD

**Issue by iammrali** [remix-run/react-router#7634]
## Version

6.0.0-beta.0

## Test Case

I create a pure component, there is no reason to execute again when i want to navigate to other page using useNavigate , I think useNavigate hook cause waste rendering on pure components, when using useHistory hook of last version(5) it doesn't have this behavior
I have a question [here ](https://stackoverflow.com/questions/64085975/why-usenavigate-cause-waste-render-in-react-router-dom-6)for more detail
Am i wrong or something changed?
https://codesandbox.io/s/dawn-wave-cy31r?file=/src/App.js

## Steps to reproduce

Create pure component
Use useNavigate hook inside it

## Expected Behavior

Navigate to other page should not cause execute of pure components when props are not changed

## Actual Behavior

Navigate to other page in pure component causes waste rendering, component execute again even when props are same

**Comment by timdorr** [comment]
`useNavigate` changes when the current location changes. It depends on it for relative navigation. Wrapping it in `memo` only prevents re-renders from parent components. If hooks within the component cause re-renders, there is nothing `memo` can do.

**Comment by iammrali** [comment]
I create this test case in v5 and use useHistory hook it doesn't have this behavior, why useNavigate isn't like that?

**Comment by benneq** [comment]
Please reopen. This causes half of my application to rerender, just because some searchParams change.

> `useNavigate` changes when the current location changes.

But the current location is only relevant for `useNavigate`'s internal state. There's no reason to recreate the navigate function each time.

**Comment by bennettdams** [comment]
_(For future readers: Workarounds shown below)_

Please consider changing this to the expected behavior. As @benneq said, this causes a lot of rerenders.
To my understanding, `useNavigate` does not need existing search params on the current location. If you need them, [you can  provide them via the hook already,](https://github.com/ReactTraining/react-router/issues/7743#issuecomment-770296462) so why not letting `useNavigate` bail out of rerendering?

## Why this is a problem
_Example of a real life use case_

In my case, I have a list of very render-intense charts. I want to click the charts to navigate to a new page for a detailed view.
Imagine each chart showing the temperatures of one day:
```tsx
// child
function DayChart({ dayData: DayData }) {
  const navigate = useNavigate();

  return (
    <div onClick={() => navigate(`/my-route/${dayData.id}`)}>
      <Chart data={dayData} />
    </div>
  );
}

// parent
function DaysList() {
  const { days } = useAPI(...)
  return days.map((day) => <DayChart dayData={day} />);
}
```
_(I'm using memoization in my code, just showing the structure here.)_

Now, `DaysList` has filters that are applied as search params. So changing any filter in `DaysList` will always rerender ALL charts, which is triggered by `useNavigate` "rerendering", which furthermore is triggered by changes in the `location` (search params). 

## Proposed solution

If `useNavigate` would ignore the search params when checking for the location, this wouldn't happen (I guess?). This cannot be done by the user (aka. developer who's using this library).

## Workarounds
There are two workarounds I came up with:

1. **Provide a wrapper function for `navigate` from the parent to the child**
You can use `useNavigate` in the parent instead of the children. This will still cause a rerender in the parent when changing search params, but this will not trigger a rerender of your children when they're memoized. For my example above:

```tsx
// child
export const DayChart = memo(function DayChart ({
  dayData,
  navigateToDay,
}: {
  dayData: DayData;
  navigateToDay: (dayId: string) => void;
}) {
  return (
    <div onClick={() => navigateToDay(dayData.dayId)>
      <Chart data={dayData} />
    </div>
  );
});

// parent
function DaysList() {
  const { days } = useAPI(...)
  const navigate = useNavigate();

  const navigateToDay = useCallback(
    (dayId: string): void => {
      navigate(`/my-route/${dayId}`);
    },
    [navigate]
  );

  return days.map((day) => <DayChart dayData={day} navigateToDay={navigateToDay} />);
}
```
BTW: I'm not really sure why `navigateToDay` does not trigger a rerender of the memoized child when it's declared in the component - I thought it will be created in the parent everytime, but somehow React knows to not rerender the child, even the `useCallback` in the parent is executed every time. Does someone know why?

2. **Use `<Navigate ... />` instead of `useNavigate`:**
If you don't know, `Navigate` is a component that, when rendered, will "execute" a `navigate`. So you could have some small boolean and conditionally render this component (in my case on button click):

```tsx
// child
function DayChart({ dayData: DayData }) {
  const navigate = useNavigate();
  const [shouldNavigate, setShouldNavigate] = useState<boolean>(false)

  return (
    <div onClick={() => setShouldNavigate(true)}>
      {shouldNavigate && <Navigate to={`/my-route/${dayData.dayId}`} />}  
      <Chart data={dayData} />
    </div>
  );
}

// parent
function DaysList() {
  const { days } = useAPI(...)
  return days.map((day) => <DayChart dayData={day} />);
}
```

**Comment by Hummel37** [comment]
Could this reopen?
For me a lot rerenders if the state changes. This is irrelevant for the useNavigate hook.
And all <Routes> components rerenders too. Do they really need the state of the current navigation?
@timdorr

**Comment by cybercoder** [comment]
same problem here, when using ```useNavigate``` there's wasted renders.

**Comment by chaance** [comment]
Re-opening for consideration. I'm not sure that we'll change this behavior but I don't want to close the door on it just yet.

**Comment by Hummel37** [comment]
I think if this PR get merged https://github.com/facebook/react/pull/20646 into react it can be used to prevent a lot of rerenders.

**Comment by SampsaKaskela** [comment]
This probably also causes issue when including navigate as dependency to useEffect hook. At least in v5 when I include the history.push as dependency, it doesn't change, like expected but now with v6 I have to include navigate to avoid warnings but the navigate keeps getting redefined which causes useEffect to run constantly. Is there any workaround for this to satisfy useEffect dependencies other than ignoring it?

**Comment by LuisDev99** [comment]
Any news on this? 

The only simple alternative I can think of is to reload the whole page when navigating to avoid two renders but there's no support for [`forceRefresh`](https://github.com/remix-run/react-router/issues/8242) in V6. I don't know how else to avoid the additional render

**Comment by aparshin** [comment]
@timdorr @chaance It seems to me that from the end-user perspective, `useNavigate` hook itself shouldn't cause re-rendering of the component. The semantic of this hook is to perform navigation. If you intentionally update the result of `useNavigate()` hook each time after location change, you implicitly suppose that the `navigate()` method will be called each time. I don't think that it's a common behavior (personally I even can't imaging such use cases). If someone wants to perform some actions based on location, he/she might use the `useLocation` hook instead.

So, if you don't change result of `useNavigate` hook each time, you'll better meet end-user expectations. Also, this will improve performance out of the box in many situations.

A separate question is how to implement this semantic at technical level. If you agree with the general idea, I can try to propose a PR (for example, something based on @RandScullard 's idea above).

**Comment by WinterWoods** [comment]
When upgrading from V5 to V6, the sub components jump, resulting in re rendering of the whole page, resulting in low performance. Is there a solution If not, you can only roll back the V5 version

**Comment by HansBrende** [comment]
`useRef` probably won't fix this issue because just the call to `useContext` (to get the location pathname) will still trigger a re-render! 

I'm guessing this is the underlying problem: https://github.com/facebook/react/issues/14110

... and can be fixed whenever react finishes their `useContextSelector` hook implementation: https://github.com/facebook/react/pull/20646

**Comment by HansBrende** [comment]
@chaance hopefully my above comment opens "the door" to a solution?

**Comment by RandScullard** [comment]
@HansBrende You're right, my useRef suggestion wouldn't prevent the extra renders because of the underlying reference to the LocationContext. Thank you for clarifying! (I think my useRef would help with issue #8349 though.)

I have one question about your suggested implementation. You say "The key to fixing this particular issue is: don't call useContext at all" but the first two lines of your function are calls to useContext. Is the key difference that these two calls are referencing context that never (or rarely) changes?

**Comment by HansBrende** [comment]
@RandScullard 
> Is the key difference that these two calls are referencing context that never (or rarely) changes?

Yes, exactly! The navigation context, for example, is set once at the top of your application by the `Router`, and would only get reset (causing all callers of `useContext(NavigationContext)` to re-render) if the `basename`, `static`, or `navigator` properties on the router change (not likely). So those `useContext` calls aren't problematic. 

`LocationContext`, on the other hand, is frequently changing so you have to watch out for those `useLocation` calls! In fact, there are other react-router hooks where it is called which could also be problematic (e.g. in `useRoutes`: **see digression below**). 

**\<digression>**
What `useRoutes` really needs is a `usePathname` so it doesn't trigger re-renders every time the query or hash changes but not the pathname! E.g.:

```javascript
function usePathname(locationOverride?: Partial<Path>) {
    const {navigator: nav} = React.useContext(UNSAFE_NavigationContext) as unknown as {navigator: History};
    const [navigator, pathname] = locationOverride ? [null, locationOverride.pathname] : [nav, nav.location.pathname];
    const [, triggerRerenderOnlyOnPathnameChange] = React.useReducer(pathnameReducer, pathname || '/');
    React.useLayoutEffect(() => navigator?.listen(triggerRerenderOnlyOnPathnameChange), [navigator]);
    return pathname || '/';
}

const pathnameReducer = (_: string, upd: Update): string => upd.location.pathname;
```

Then in `useRoutes`:
```javascript
// let locationFromContext = useLocation(); // <-- Don't do this!
// let location;
// if (locationArg) {
//    location = typeof locationArg === "string" ? parsePath(locationArg) : locationArg;
//  } else {
//    location = locationFromContext;
//  }
//  let pathname = location.pathname || "/";

// INSTEAD:
let pathname = usePathname(typeof locationArg === "string" ? parsePath(locationArg) : locationArg)
```
**\</digression>**

**Comment by HansBrende** [comment]
Ok, so, based on my "EDIT" to my above comment (and thanks to @RandScullard for getting my gears turning), to really fix this problem you will need to remove _both_ the `useLocation()` and `useContext(RouteContext)` calls. It is **really** easy to shoot yourself in the foot with `useContext`! The removal of `useLocation()` is quite easy, as I've already shown. Removing `useContext(RouteContext)` is a little more involved because it requires tweaking `_renderMatches()`. Here's an updated fix:

```javascript
export function useNavigate(): NavigateFunction {
  let { basename, navigator } = React.useContext(NavigationContext);
  // let { matches } = React.useContext(RouteContext); // <-- DO NOT DO THIS!!!!!
  // let { pathname: locationPathname } = useLocation(); // <-- DO NOT DO THIS!!!!!
  // Instead, we need to use Contexts that do not get updated very often, such as the following:
  const routeContextRef = React.useContext(NonRenderingRouteContext);
  ...
  return React.useCallback(
    (to: To | number, options: NavigateOptions = {}) => {
      ...
      // Look up the current pathname AND ROUTE CONTEXT *at call-time* rather than current behavior of:
      //   1. re-rendering on every location change (incl. query, hash, etc.) (BAD!)
      //   2. creating a new navigate function on every pathname change (BAD!)
      let path = resolveTo(
        to,
        routeContextRef.current.matches.map(match => match.pathnameBase),
        (navigator as History).location.pathname
      );
      ...
      navigator.doSomethingWith(path); // ta-da!
    },
    [basename, navigator, routeContextRef]
  );
}

// Where the "NonRenderingRouteContext" is provided, for example, as follows:

const NonRenderingRouteContext = React.createContext<{readonly current: RouteContextObject}>({
  current: {
    outlet: null,
    matches: []
  }
});

function RouteContextProvider({value}: {value: RouteContextObject}) {
  const ref = React.useRef(value);
  React.useLayoutEffect(() => void (ref.current = value));
  const match = value.matches[value.matches.length - 1];
  return (
      <NonRenderingRouteContext.Provider value={ref}>
        <RouteContext.Provider value={value}>
          {match.route.element !== undefined ? match.route.element : <Outlet />}
        </RouteContext.Provider>
      </NonRenderingRouteContext.Provider>
  );
}

function _renderMatches(matches: RouteMatch[] | null, parentMatches: RouteMatch[] = []): React.ReactElement | null {
  if (matches == null) return null;
  return matches.reduceRight((outlet, match, index) => {
    return (
      <RouteContextProvider  // Simply swap out RouteContext.Provider for RouteContextProvider here!
        value={{
          outlet,
          matches: parentMatches.concat(matches.slice(0, index + 1))
        }}
      />
    );
  }, null);
}
```

**Comment by ryanflorence** [comment]
Is anybody measuring performance issues here or just console logging "rerender!" and stressing out?

I'd be interested to see an example where this actually affects the user experience.

The fact is that `navigate` depends on `location`, so it changes with it.

**Comment by ryanflorence** [comment]
Just reviewed some of the code samples here.

Refs shouldn't be assigned during render, only in effects or event handlers. Maybe the React team has changed their position on this recently, but mutating refs while rendering can cause "tearing" in concurrent mode/suspense. We're just following the rules here: pass dependencies to effects.

Since navigate isn't "rendered", it is unfortunate that it's dependency on `location` causes re-renders.

If you've got actual performance problems (not just console logging "rerender!") the simplest solution I can think of right now is to call `useNavigate` and the top and make your own context provider with it--but remember not to use any relative navigation.

```js
let AbsoluteNavigate = React.createContext();

function Root() {
  let [navigate] = useState(useNavigate())
  return (
    <AbsoluteNavigate.Provider value={navigate}>
      <Routes>
        {/*...*/}
      </Routes>
    <AbsoluteNavigate.Provider>
  )
)
```

**Comment by LuisDev99** [comment]
@ryanflorence My issue: Whenever the component mounts, I make a HTTP request. Because of the double render issue, the component is making two HTTP requests, and for my particular case, we are logging every HTTP request made to our service, but since the component is making an extra request, we are logging two events when there should've been just one.  

That's why the double render in my use case is a big problem. Surely we can implement some sort of caching to prevent additional requests, but I'm just trying to give one real example.

**Comment by deleteme** [comment]
@ryanflorence This isn't just an optimization problem. Another problem that caused user-facing bugs while upgrading from v5 to v6, was that a `navigate` call within a `useEffect` hook would incur extra unexpected calls.

Here's the scenario:

We have a field with an onComplete handler that, when triggered, navigates to a new route.

1. A custom input component receives an `onComplete` prop.
2. The input value may be set and change from multiple different ways, so some special logic isn't suitable across various event listeners and instead centralized in a reducer.
3. A useEffect hook receives `onComplete` and the input state as dependencies. When the hook determines the input state value to be complete, `onComplete` is called.

```
useEffect(() => {
  const isComplete = /* something with the state */
  if (isComplete) {
    onComplete(state);
  }
}, [state, onComplete]);
```

4. This custom input component is mounted within a page that has a couple routes and also uses search params. The component will remain mounted throughout all of these routes.
5. `onComplete` is memoized with `useCallback`, and calls `navigate` with one of those route urls.
6. Bug: When the input is complete, the location changes, then `navigate` is redefined, so `onComplete` is redefined, and because it's a new function, triggers the `useEffect` `onComplete` callback an extra time.

A work-around was to decouple the `onComplete` callback from that effect, then dispatch and listen to a custom event:

```
  // announce that the input is complete via a custom event
  useEffect(() => {
    const element = elementRef.current;
    if (isComplete && element) {
      element.dispatchEvent(
        new CustomEvent(ONCOMPLETE, {
          bubbles: true,
          detail: { template, valueString },
        })
      );
    }
  }, [template, valueString, isComplete, elementRef]);

  // call an onComplete callback
  useEffect(() => {
    const element = elementRef.current;
    if (onComplete && element) {
      const handler = (e) => onComplete(e.detail);
      element.addEventListener(ONCOMPLETE, handler);
      return () => {
        element.removeEventListener(ONCOMPLETE, handler);
      };
    }
  }, [onComplete, elementRef]);
```

**Comment by RandScullard** [comment]
@ryanflorence I agree with @deleteme, I am really more concerned with effects being triggered. Additional renders don't create a problem in my use case.

I still think the effect problem could be solved as in https://github.com/remix-run/react-router/issues/7634#issuecomment-1006997643. But good point about setting the ref in an effect, not inline in the render. (Full disclosure: I have not tested this.)

**Comment by Brammm** [comment]
Just chiming in, came to this issue after running into a race condition. I was optimistically updating state and navigating away from the page to an updated page slug, but because `navigate` isn't memoized, my app would load it's data again, which wasn't updated yet in the db.

**Comment by Hummel37** [comment]
What do you think about a hook like useAbsoluteNavigate() or so. The hook would be just like useNavigate() but would not support relative paths. Then you wouldn't need useLocation in the hook and would save some rerenderer.

**Comment by Hummel37** [comment]
Or can we just use window.location.pathname instead of useLocation().pathname?

**Comment by flexdinesh** [comment]
To anyone who's looking for an interim solution to prevent re-rendering of components using `useNavigate()` and `useLocation()` I've got something I've been using for a while now. 

```jsx
import React from 'react';
import {useNavigate as useNavigateOriginal, useLocation as useLocationOriginal} from 'react-router-dom';
import type {Location, NavigateFunction} from 'react-router-dom';

type RouterUtilsContextType = {
  navigateRef: React.MutableRefObject<NavigateFunction> | null;
  locationRef: React.MutableRefObject<Location> | null;
};
const RouterUtilsContext = React.createContext<RouterUtilsContextType>({
  navigateRef: null,
  locationRef: null,
});

/*
  react-router uses one big context to send changes down the react tree.
  So every route or query param change will re-render the context which will in-turn re-render 
  all the hooks subscribed to react-router context - useNavigate(), useLocation().
  This prevents us from using these hooks as utilities to get latest location or query param value 
  in a component since all the components using these hooks will re-render in addition to the 
  entire Route component re-rendering - which is not ideal.

  With this RouterUtilsContext - we tank the updates from react-router context and
  drill down navigate and location from a separate context.
  This will prevent re-render of consumer components of these hooks for every route change
  and allow using these hooks as utilities instead of context subscribers
*/
const RouterUtils: React.FC = ({children}) => {
  const navigate = useNavigateOriginal();
  const location = useLocationOriginal();

  // useRef retains object reference between re-renders
  const navigateRef = React.useRef(navigate);
  const locationRef = React.useRef(location);

  navigateRef.current = navigate;
  locationRef.current = location;

  // contextValue never changes between re-renders since refs don't change between re-renders
  const contextValue = React.useMemo(() => {
    return {navigateRef, locationRef};
  }, [locationRef, navigateRef]);

  // since contextValue never changes between re-renders, components/hooks using this context
  // won't re-render when router context updates
  return <RouterUtilsContext.Provider value={contextValue}>{children}</RouterUtilsContext.Provider>;
};

/* 
  Please be aware: when the url changes - this hook will NOT re-render 
  Only use it as a utility to push url changes into Router history
  which will then re-render the whole route component.
  Eg. const navigate = useNavigateNoUpdates();
*/
export const useNavigateNoUpdates = () => {
  const {navigateRef} = React.useContext(RouterUtilsContext);
  if (navigateRef === null) {
    throw new Error(
      'RouterUtils context should be added to the React tree right below BrowserRouter for useNavigateNoUpdates hook to work. If you need router in tests or stories, please use WrappedMemoryRouter utility.',
    );
  }
  return navigateRef.current;
};

/* 
  Please be aware: when the url changes - this hook will NOT re-render 
  Only use it as a utility to get latest location object.
  Eg. const location = useLocationNoUpdates();
*/
export const useLocationNoUpdates = () => {
  const {locationRef} = React.useContext(RouterUtilsContext);
  if (locationRef === null) {
    throw new Error(
      'RouterUtils context should be added to the React tree right below BrowserRouter for useLocationNoUpdates hook to work. If you need router in tests or stories, please use WrappedMemoryRouter utility.',
    );
  }
  return locationRef.current;
};

/* 
  Please be aware: when the query params change - this hook will NOT re-render. 
  Only use it as a utility to get latest query params value.
  Eg. const params = useQueryParamsNoUpdates();
  const sidebarGoalId = params['sidebar_goal_id'];
*/
export const useQueryParamsNoUpdates = () => {
  const {search} = useLocationNoUpdates();

  const queryParams = React.useMemo(() => {
    const urlSearchParams = new URLSearchParams(search);
    const params = Object.fromEntries(urlSearchParams.entries());
    return params;
  }, [search]);

  return queryParams;
};

/* 
  Please be aware: when the query param changes - this hook will NOT re-render. 
  Only use it as a utility to get latest query param value.
  Eg. const sidebarGoalId = useQueryParamNoUpdates('sidebar_goal_id');
*/
export const useQueryParamNoUpdates = (name: QueryParams) => {
  const params = useQueryParamsNoUpdates();

  if (!name) {
    throw new Error(`useQueryParam name arg cannot be empty — name: ${name}`);
  }

  return params[name];
};

export default RouterUtils;

```

And you add this Provider right below the Router provider in the component tree

```jsx
return (
    ...
      <Router>
        <RouterUtils>
          ...
        </RouterUtils>
      </Router>
    ...
  );
```

**Comment by houmark** [comment]
@ryanflorence If the goal was to slow down user adoption of v6 and have people look for alternative routers, then that goal has certainly been reached with a breaking approach like this one.

The documentation on migration from v5 to v6 in relation to `push` and `navigate` goes like this:

> In v6, this app should be rewritten to use the navigate API. Most of the time this means changing useHistory to useNavigate and changing the history.push or history.replace callsite.

Ref: https://reactrouter.com/docs/en/v6/upgrading/v5#use-usenavigate-instead-of-usehistory

Nowhere is it mentioned that the fingerprint of the `useNavigate` hook is changing on every function call. This leads to hard-to-fix unexpected bugs for many solutions in the wild.

In our case, and in many other cases I've read here and elsewhere, a `useEffect` does navigation to the same component. That's not a wrong pattern or use case, but there is no simple alternative or solution to solving that. Sure, we can start wrapping the "hack" you drafted above, but that is not what you'd expect when migrating from v5 to v6 — and if there are major difference you'd expect it would be highlighted and reasonable workarounds would be exampled. We have had a fairly simple migration up until we realized that we had several cases where this breaks the entire component navigation and that there was no simple way to fix it.

We can now either spend hours on doing some hacky workarounds, rethink our entire component approach (hard without major refactor) or stash our migration and once again put the v6 migration on the backburner. Unfortunately, React 18 seems to have issues with StrictMode with React Router v5 and while the rest of our app is ready to upgrade to React 18, upgrading to v6 together with React 18 made the most sense.

I echo other simple solutions above where `useNavigate` internally can keep track of the relative path, but it avoids changing on navigate. I understand it may not be "cleaner" to do it that way (or how React wants you to), but it's certainly a much friendlier backward compatibility approach that will have fewer users pulling their hair.

**Comment by houmark** [comment]
The solution outlined by @RandScullard in https://github.com/remix-run/react-router/issues/7634#issuecomment-1006997643 seems like a very simple and low-risk change, that will keep current functionality while also introducing better backward compatibility without breaking any exitisting adoptions to `navigate` — and I think using useRef is also an acceptable React approach.

There is literally no upside to the function changes that I can identify and also no negative side effects from the hook not returning the locationPathname, as one can fetch that using the `useLocation()` hook anyways.

Considering that, would you accept a PR that does that change @ryanflorence with the purpose of easing adoption? There should not be a lot of work to do that, but I would still like to wager your interest in such a change before working on a PR.

**Comment by chaance** [comment]
This conversation is getting a bit heated, so let's please bring it down a notch. Please remember this is free software maintained by people who are trying to help in good faith.

Ryan stated above that he'd consider a PR that makes a change as long as it's safe and follow all the aforementioned rules, so if this is a priority for you please let us know when that PR is ready and we'll take a look as soon as we can. In the mean time, there's always [`patch-package`](https://www.npmjs.com/package/patch-package)!

**Comment by houmark** [comment]
I wouldn't call it heated, and if it came across that way, then please understand that was not my intention. My intention was to bring attention to the fact that this was not clear in any way in the docs, and I am surprised by the decision to make such a change without really considering the consequences. If it was a bug that slipped through, then it's another story, but then I would assume it would have been fixed since this was reported in 6.0.0-beta.0 months ago, especially considering that the fix is pretty trivial.

I think several people in this thread, myself included, were a tiny bit triggered by Ryan's initial responses where he downgraded the issue and if that the priority was low, when this is maybe the one most breaking change or upgrade frustration developers may encounter in v5 to v6 migration — and situation that has no reasonable solution.

I asked for confirmation of the approach for a fix, as Ryan has also indicated his stance on various solutions and the one I am voting for and would attempt in a PR did not seem to be his favorite approach by reading his comments, so maybe it's not within his rules.

Living life with `patch-package` is not a life worth living imho, so I will try to find time to submit a PR unless someone else has the interest to do so in the very near future.

**Comment by jamestalton** [comment]
The issue that the useNavigate changes every time location changes has caused useEffect problems for myself several times.

My workaround:
```
function useNav() {
    const navigate = useNavigate()
    const navigateRef = useRef({ navigate })
    useEffect(() => {
        navigateRef.current.navigate = navigate
    }, [navigate])
    return useCallback((location: string) => {
        navigateRef.current.navigate(location)
    }, [])
}
```

**Comment by HansBrende** [comment]
@houmark : only problem with Rand's solution is that it only fixes half the problem, the unstable navigate function. The "waste rendering" problem which is the title of this issue is actually still present.

**Comment by willd-i2o** [comment]
This issue is blocking our migration to v6, we'll be stuck on v5 or switching to an alternative routing solution if this remains unaddressed.

**Comment by chaance** [comment]
I'm keeping this issue open but locking further comment. As already stated, PRs that address this are welcome. If this is blocking your team we encourage contributions.

**Comment by ryanflorence** [comment]
This issue can be solved with the new routers in 6.4.

```jsx
import { createBrowserRouter } from "react-router-dom";
const router = createBrowserRouter([ /* ... routes ... */ ]):
export { router }
```

```jsx
import { router } from "./router";

// call this anywhere you want in your app,
// it's outside of the react render cycle and will
// never change
router.navigate("/anywhere/you/want");
```

With the new routers/histories in 6.4, I think `router.navigate(relativePath, routeId?)` could work without needing any component code to figure out the relative paths, and therefore keep the identity of the function the same. Unlike the `history` objects before it:

- It knows the route tree, so we can internally pass the route ID to know where in the tree we need to construct the relative URL
- It holds the state, and therefore knows the current location

Previously, that information was in the React render tree, and that's why the stuff was passed to `useCallback`, changing the identity of the function.

@brophdawg11 do you think this is possible now?

**Comment by imnotjames** [comment]
Wouldn't you still need `history` in `useNavigate` because of the `to: number` for `navigate(-1)`, etc?

**Comment by brophdawg11** [comment]
I added a Proposal for this behavior in https://github.com/remix-run/react-router/discussions/9588

**Comment by MacgyverMartins** [comment]
Is there any way to programmatically navigate without using useNagivate on V6? 
useNagivate causes a lot of renders in my code :\

**Comment by MacgyverMartins** [comment]
Thanks, @flexdinesh. your context solution helped me a lot. 
After moving to v6 our app started to have performance problems because of the renders caused by useNavigate

**Comment by hichemfantar** [comment]
> This issue can be solved with the new routers in 6.4.
> 
> ```js
> import { createBrowserRouter } from "react-router-dom";
> const router = createBrowserRouter([ /* ... routes ... */ ]):
> export { router }
> ```
> 
> ```js
> import { router } from "./router";
> 
> // call this anywhere you want in your app,
> // it's outside of the react render cycle and will
> // never change
> router.navigate("/anywhere/you/want");
> ```
> 
> I think we can also get away with keeping the identity for the `navigate` that `useNavigate` returns too.
> 
> With the new routers/histories in 6.4, I think `router.navigate(relativePath, routeId?)` could work without needing any component code to figure out the relative paths. Therefore when returned from `useNavigate` we could keep the identity of the function the same and just wrap `router.navigate` with the ID inside like `router.navigate(path, routeId)`.
> 
> Unlike the `history` objects before it:
> 
> * It knows the route tree, so we can internally pass the route ID to know where in the tree we need to construct the relative URL
> * It holds the state, and therefore knows the current location
> 
> Previously, that information was in the React render tree, and that's why the stuff was passed to `useCallback`, changing the identity of the function.
> 
> @brophdawg11 do you think this is possible now?

Thanks. Using the exported router to navigate seems to solve the multiple rendering problem.
Any plans to make this work with the useNavigate hook?

**Comment by Mikilll94** [comment]
@ryanflorence 
That's awesome. Can you put more priority to this? Currently this unnecessary rendering is a big issue of react-router. Fixing this would be a big value for the users.

**Comment by kellengreen** [comment]
Am I correct to assume that `6.4` does not address this for developers using `<BrowserRouter>`, `<Route>`, and `useNavigate()`?

**Comment by brophdawg11** [comment]
#10336 fixes this when using a `<RouterProvider>` and will be available in the next release

**Comment by brophdawg11** [comment]
Reopening until this is released to npm

**Comment by kellengreen** [comment]
> @kellengreen That's correct - `useNavigate` has to change when the `location` changes because it's dependent on `useLocation` for it's relative-routing logic. With the introduction of `RouterProvider` `useNavigate` only needs to know the route ID it was called from, allowing it to be stable across location changes.

Is there plan to fix this with say a `useNavigateAbsolute` hook? I would assume`BrowserRouter` is still the most popular implementation method.

**Comment by kellengreen** [comment]
Thanks @brophdawg11 I'll take a look.

**Comment by brophdawg11** [comment]
Aligning with Remix here - will close this issue with `awaiting release`.  This will be available when React Router 6.11 is released.

**Comment by braincore** [comment]
@brophdawg11 Gave a quick try using `6.11.0` and your minimal `RouterProvider` approach. I'm still seeing `navigate()` trigger components that use the `useNavigate()` hook.

**Comment by hichemfantar** [comment]
@braincore can you please provide a code sandbox example.

**Comment by braincore** [comment]
@hichemfantar Here's a sandbox: https://codesandbox.io/s/admiring-dan-uvyqeg?file=/src/App.js

The `<Frame>` components calls `useNavigate()`. Meanwhile, its `children` routes navigate between each other (changing the location), which triggers unnecessary renders of `<Frame>`. See the console for a printout every time `Frame` is rendered.

**Comment by brophdawg11** [comment]
@braincore I think there's some confusion with re-renders versus navigate identity stability, so per [this comment](https://github.com/remix-run/react-router/issues/7634#issuecomment-1306650156), what we did here was make the identity of the `navigate` function stable.  You can see this in https://codesandbox.io/s/kind-wilson-xgn211?file=/src/App.js.

Most of the actual issues reported were regards to the `navigate` identity changing:

> a navigate call within a useEffect hook would incur extra unexpected calls

> I am really more concerned with effects being triggered. Additional renders don't create a problem in my use case.

> a useEffect does navigation to the same component

> The issue that useNavigate changes every time location changes has caused useEffect problems for myself several times.

But, `useNavigate` it still reads from the `RouteContext` to determine the current route id, and that changes across navigations so it will re-render, but it won't trigger effects depending on `navigate`.  

If you want to optimize further and avoid re-renders, there are two suggested solutions above to either use a [context-provided `navigate`](https://github.com/remix-run/react-router/issues/7634#issuecomment-1017572625) or [navigate directly via the `router`](https://github.com/remix-run/react-router/issues/7634#issuecomment-1306650156).

**Comment by braincore** [comment]
@brophdawg11 Thanks for the clarification. I've given this another try, and the stable navigate is sufficient for my use case. I'm able to reduce renders using typical react memoization strategies.  Thanks!

**Comment by smitpatelx** [comment]
Is there any way I could wrap `RouterProvider` around components?
```tsx
<RouterProvider router={router}>
  <RefreshToken />
  <Toaster />
</RouterProvider>
```
Typescript error: `Property 'children' does not exist on type 'IntrinsicAttributes & RouterProviderProps'`.

#### Purpose:
I want to run some function on interval and which also uses `useNavigate` hook. And that function should run on whole app.

#### Issue:
- RefreshToken component does not function properly. Not running useEffect.

**Comment by brophdawg11** [comment]
Put those components in a root [layout route](https://reactrouter.com/en/main/start/concepts#layout-routes) around your whole app:

```
let router = createBrowserRouter([{
  path: '',
  Component() {
    return (
      <>
        <RefreshToken />
        <Toaster />
        <Outlet />
      </>
    );
  },
  children: [/* app routes here */],
});
```

**Comment by cbreezier** [comment]
I'm late to the party, but I'm still not completely satisfied with the current fix.

Can someone from the team (@brophdawg11 @ryanflorence) explain why the `useNavigate` hook can't be changed to read the pathname at runtime (eg, as per https://github.com/remix-run/react-router/issues/7634#issuecomment-1013806084) instead of using `useLocation`?

**Comment by HansBrende** [comment]
@process0 from what I recall of TanStack Router, it was essentially a carbon copy of react-router with extra features baked in... this may have changed since I last looked at it though. I've been using [wouter](https://github.com/molefrog/wouter) + a custom monkeypatch to support nested routing since that library is a little less opinionated, lighter weight, and easier to grok. Importantly, its usage of react contexts is safe (i.e. doesn't re-render your whole app).

**Comment by cabello** [comment]
What I noticed on `v6.14.1` is that I tried moving a component that received children to use Outlet for example:

```
function NavbarLayout({ children })  {
  return `<div>{children}<Navbar /><div>`
}
```

To:

```
function NavbarLayout()  {
  return `<div><Outlet /><Navbar /><div>`
}
```

And all of the sudden my Navbar was blinking like mounting and remounting, I am still debugging, checking if this is only because of Strict mode in development or something.

I suspect it's related to this issue because I use `useLocation` (mentioned in this issue), to decide which item to highlight in the navbar.

**Comment by depoulo** [comment]
After reading all of this, I'm still unsure about what would be the best practice to solve the following:

I have a paginated list view. The current page is indicated by a URL search parameter (`?page=n`), with a fallback to being on page 1 when no parameter is provided. Now, when on page 2 and clicking a list entry, the app navigates to the according details view (via a react-router link). But before that happens, the list view rerenders, only that now it displays page 1, resulting in an unpleasant flash. The rerender, and thus the flash, disappear if I remove both `useNavigate()` and `useSearchParams()`, which are used in the pager component.

**Comment by bioshazard** [comment]
Just ran into this myself. Simply adding the line `const location = useLocation()` to the top of my component containing `<Routes>` causes the entire application to re-render any time a link is clicked. And its not a performance issue so much as it looks terrible because of a flicker from re-rendering lazy loaded components.

**Comment by karlpablo** [comment]
Here's my working solution using `useRef`. It doesn't need `navigate` in the dependency array anymore.

```
const navigate = useRef(useNavigate())

useEffect(() => {
  connect()
  navigate.current('/')
  return () => disconnect()
}, [])
```

I'm on v6.23.1, btw.

**Comment by BenRauzi** [comment]
Would be great @kellengreen

**Comment by enjikaka** [comment]
I've tried both useNavigate, rendering `<Navigate>`, providing the `NavigateFunction` through a React Context and directly calling Router.navigate in my codebase now. All these variants trigger the loaderData method on the current route before navigating to the `To` route. Clicking anchors hooked up with `useLinkClickHandler` is the only thing that does not trigger this behaviour.

**Comment by ingjuanbejarano** [comment]
I'm am working with `useNavigate` and `Link` components and I'm facing the same issue, the origin view is rendering everything once more before going into destination and that is causing performance issues especially in the INP, is any viable solution yet to this issue? before thinking seriously to downgrade to v5.
thanks.

**Comment by sahilanguralla** [comment]
A Redux-like location selector function would be helpful here.

```
const isMyPath = useLocationSelector((location) => location.pathname.includes('/my-path'));
```
And re-rendering can be avoided if the function's return value has not changed.

**Comment by mironenkod** [comment]
I Install the `minimal` template from [https://github.com/remix-run/react-router-templates](react-router-templates)
Run npm install
Modify Home
`export default function Home({ loaderData, ...args }: Route.ComponentProps) {
  let loc = useLocation();
  console.log("Home", loc.search);
  return (
    <div className="text-center p-4">
      <Link to={"/?p=" + Math.random().toString().slice(4,7)}>Home</Link>
      <h1 className="text-2xl">Hello, {loaderData.name}</h1>
      <a
        className="block mt-2 text-blue-500 underline hover:text-blue-600"
        href="https://reactrouter.com/docs"
      >
        React Router Docs
      </a>
    </div>
  );
}`

That's it, i don't change anything else in the minimal template.
Run npm run dev.
Go to the homepage, everything is fine.
Click the link, and we see that the current page re-renders first before the transition happens.
Only after that, the new page renders.

![Image](https://github.com/user-attachments/assets/06c5b0b2-90c3-4a06-9685-bedacb6e4ecd)

Before transitioning to another page, the entire current page just re-renders.
Is this normal behavior?

**Comment by AntonioLuisME** [comment]
We're also experiencing this issue. 5 years and no fix?!

**Comment by ghost** [comment]
ngl the unmentioned in this issue trouble is when you are using navlink/link, which under the hood is probably, high likely even uses usenavigate, therefore any sort of components, even if they are static and should get optimized or even if they are memoized, get rerendered as well, pure autism, you might think lets use usenavigate and get rid of link/navlink which is simply a wrapper on top of a <a> tag, but then we get introduced to the bug described in this issue

> Navigate to other page in pure component causes waste rendering, component execute again even when props are same

probably happens due to the change of "matches" array inside of props

**Comment by nick-delirium** [comment]
Just spent entire day debugging our application only to find out its this bug still not being fixed despite issue being closed. Why do I need to create fake location object, and my own useNavigate implementation together with using UNSAFE_NavigationContext? :/

**Comment by nicubarbaros** [comment]
@nick-delirium have you found an workaround?
